Multistep proposals

AloeareV opened the pull request.

AloeareV committed:
empty test

AloeareV committed:
test shows that we can create a multistep proposal

AloeareV committed:
wip

AloeareV committed:
fix merge conflicts and finish new_ephemeral_address

AloeareV committed:
Merge branch 'unify_keys' into multistep_proposals

AloeareV committed:
wip

AloeareV commented on libtonode-tests/tests/concrete.rs:
```diff
@@ -159,6 +167,24 @@ mod fast {
                 && vt.recipient_address() == Some(ZENNIES_FOR_ZINGO_REGTEST_ADDRESS)));
     }
 
+    #[tokio::test]
```
Here

AloeareV committed:
wip

AloeareV committed:
Merge remote-tracking branch 'origin/dev' into multistep_proposals

AloeareV committed:
novel test failure

zancas committed:
follow keys::NYM convention, in place elsewhere in this mod

zancas committed:
payments --> payment, and dbg to silence warning

zancas committed:
strat --> tex_strat

zancas commented on zingolib/src/wallet/keys/legacy.rs:
```diff
@@ -169,13 +169,12 @@ pub(crate) fn legacy_sks_to_usk(
     external_pubkey: &AccountPubKey,
     address_index: NonHardenedChildIndex,
-) -> Result<TransparentAddress, String> {
+) -> Result<TransparentAddress, bip32::Error> {
```
I am cautious around Endianness flips between [B/Z]IP32s.

AloeareV committed:
implement get_known_ephemeral_addresses, test now passes

AloeareV committed:
add comment about rename

AloeareV committed:
some cleanup

AloeareV commented:
I would like to change the test to confirm that the receiver actually receives the funds sent to the tex.

zancas committed:
apply lint

zancas committed:
simplify

zancas commented:
Running a local sanity check.

zancas committed:
Merge branch 'dev' into multistep_proposals

AloeareV committed:
start on syncing ephemeral transparent funds

AloeareV committed:
start on test of reciept to tex

zancas committed:
use a valid extended pubkey in the new_with_witness_trees_address_free test helper

zancas committed:
const

zancas committed:
use texregtest

zancas committed:
uncomment failing assertion

zancas committed:
Merge branch 'dev' into multistep_proposals

zancas committed:
add doc-comment to get_external_taddrs

zancas committed:
match on status on handle_receipt_to_ephemeral_taddr

zancas committed:
same behavior as for non-ephemeral (so far)

zancas committed:
revert addition of dbgs

zancas committed:
add doc-comment to account_for_transparent_receipts

zancas committed:
add doc-comment to record_ephem_taddr_receipt

zancas committed:
record all taddr-receipts in transaction records the same way

zancas committed:
use value transfers

zancas committed:
begin defining a wrapped SentValueTransfer enum to properly type value transfers

Oscar-Pepper commented on zingolib/src/wallet/transaction_context.rs:
```diff
@@ -216,40 +209,69 @@ mod decrypt_transaction {
+        /// A receipt of funds has been detected at a ZIP320 "ephemeral" return
+        /// address for a Transparent-Source-Only encoded "TEX" address.
+        /// This method records that receipt in therelevant receiving
+        /// TransactionRecord in the TransactionRecordsById database.
```
this comment seems to only refer to ephemeral but it is also called when the output_taddr is found in the taddrs_set (external)

zancas replied:
Ready for resolution.

Oscar-Pepper commented on zingolib/src/wallet/tx_map/spending_data.rs:
```diff
@@ -15,13 +15,16 @@ pub(crate) struct SpendingData {
     #[getset(get = "pub(crate)", get_mut = "pub(crate)")]
     cached_raw_transactions: HashMap<TxId, Vec<u8>>,
+    #[getset(get = "pub(crate)", get_mut = "pub(crate)")]
+    transparent_ephemeral_ivk: EphemeralIvk,
```
can you give some context to what problem are we trying to solve by having keys held anywhere other than UnifiedKeyStore?

AloeareV replied:
We don't want to our WalletRead/WalletWrite implementers to need spending keys. One reason is because we want a hardware wallet to be able to call the transaction creation step, without zingo holding the keys, and then passing the transaction to the hardware wallet to sign.

There are a couple places where we need viewing keys for walletread/walletwrite, so we copy those keys into the TxMap.

Oscar-Pepper replied:
this leads me to believe that walletwrite/ walletread is implemented on the wrong struct, which was my suspicion in the past. we can fix this later!

Oscar-Pepper commented on zingolib/src/wallet/send.rs:
```diff
@@ -143,11 +144,22 @@ impl LightWallet {
+                    .await
+            }
+            2 /*TODO: check that this is actually tex*/ => self.create_transaction_helper(sapling_prover, unified_spend_key, proposal).await,
+            _ => panic!("TODO: error handle"),
```
can we change the BuildTransactionError::ExchangeAddressesNotSupported variant to something else that we put here?

zancas replied:
Ready for resolution.

Oscar-Pepper commented on zingolib/src/wallet/keys/unified.rs:
```diff
@@ -695,6 +627,15 @@ impl WalletCapability {
         }
     }
+
+    pub(crate) fn ephemeral_ivk(
```
I think its a mistake to have key related helper functions attached to WalletCapability, I think all key related stuff should stay contained in UnifiedKeyStore. UnifiedKeyStore has a TryFrom for AccountPubKey so I dont see why this helper is neccessary but I think it should at least be on UnifiedKeyStore

zancas replied:
Resolved by preceding comment.

AloeareV replied:
This function goes WalletCapability->AccountPubKey->EphemeralIvk. I was calling `AccountPubKey::try_from(wallet_capability().unified_key_store()).ephemeral_ivk()` in enough places I wanted to make a helper function for it.

Turning it into a unified_key_store method means that in all the places it's used, we'd need to add `.unified_key_store()` first...which isn't a very large cost, but this function exists to combine 3-ish lines of code to 1, and going down to 2 lines of code defeats half the point. 

What makes a helper 'necessary'? If you're worried about cluttering the `unified_key_store` impls, would moving it to be a free function instead of a method solve that?

Oscar-Pepper replied:
ok thats fine. I think i am anticipating that the keystore might move eventually so its best to keep it contained but we can cross that bridge later.

Oscar-Pepper commented on zingolib/src/wallet/keys/unified.rs:
```diff
@@ -421,31 +400,9 @@ impl WalletCapability {
-        } else {
-            None
-        };
+        let transparent_receiver = self
```
why have we removed the desired_receivers check and also the fvk generation so its unified with the logic of sapling and orchard? I did this work on my last PR and its much cleaner and important to prevent bugs with UFVKs without transparent viewing keys

AloeareV replied:
Oh lord...I think this may have been lost when I was trying to fix merge conflicts.

zancas replied:
Is this resolved to your satisfaction @Oscar-Pepper ?

AloeareV replied:
Yes, I think on one version it was moved, and when I resolved conflicts it was now in neither location...added back, good catch.

zancas replied:
@AloeareV has re-added it.

AloeareV committed:
unexpected error

zancas committed:
begin updating doc-comment

zancas committed:
work on doc-comment

todo actually create new memos

add comment

add comment

TEX is mutually exclusive with shield

write ephemeral address index to memo

AloeareV committed:
cached_raw transactions HashMap -> Vec

AloeareV committed:
handle uas in version 1 memo

zancas committed:
fix doc-comment, and refer to non-self as 'foreign'

zancas committed:
migrate zingomemo handling by TransactionContext into separate mod

zancas committed:
moving the if let out the iteration over the us loop is cleaner

zancas committed:
move if let to outer fn

zancas committed:
start writing texes from memo index into out_metadata.recipient_address

zancas committed:
start defining ephemeral mod for WalletCapability

zancas committed:
move more impl into ephemeral

zancas committed:
define an appropriate error

zancas committed:
add error handling to tex address extraction

zancas committed:
make previously invisible errors explicitly expected

zancas committed:
allow dead code

zancas committed:
start infallible construction of ephemeral addresses

zancas committed:
iterate over missing ephemeral key indices

zancas committed:
check whether we can remove transaction from parameters of handles_texes

zancas committed:
add new ephemeral key

zancas committed:
some doc-comment refactor

AloeareV committed:
Merge remote-tracking branch 'origin/dev' into multistep_proposals

AloeareV committed:
lock update after merge

AloeareV committed:
fix new_ephemeral_address -> new_persistant_ephemeral_address, now wraps ephemeral_address, both made callable from more scopes

handle the index equal to total keys case

fix test errors

update function name

remove sanity check

note that this doesn't handle upside down sync

use wrapper

add TODO, for removing transparent_child_ephemeral_addresses

update parameter name

wallet.blocks --> wallet.last_100_blocks

update uniquely called fn parameter to reflect caller ident

mysterious wtx --> prev_tx_record

also look up ephemeral taddr transactions

move taddr_set bind

finish draft of helperization

not a &mut on a Copy

put the set of taddrs together

assume we always care about all taddrs

test code refactor

texify

comment

start stubbing out tex detection

fix whitespace and imports

cthulhu hates trailing white space

apply lints

ready for sanity tests

learn from clippy

grab the tex address as a String

clean up lints

remove cascade of total_transparent_balue_spent

working

AloeareV committed:
fix order of transaction summaries

AloeareV committed:
1 -> 2

zancas committed:
correct doc-comment

fluidvanadium committed:
group test functionality

AloeareV committed:
add back removed check

AloeareV committed:
Merge remote-tracking branch 'aloe/multistep_proposals' into multistep_proposals_rebased

fluidvanadium committed:
upgrade testname

fluidvanadium committed:
Merge remote-tracking branch 'aloe/multistep_proposals' into HEAD

fluidvanadium committed:
upgrade comment

fluidvanadium committed:
fix spelling

fluidvanadium committed:
group test functionality
